File: Lectures & Talks/Day 3/Examples/mpi_collective.py
#!/usr/bin/env python

import math
import numpy as np
import mpi4py; mpi4py.rc.recv_mprobe = False
from mpi4py import MPI

N = 100; n = 0; root = 0
result = np.zeros(1, dtype='i')
final_result = np.zeros(1, dtype='i')

comm = MPI.COMM_WORLD
numtasks = comm.Get_size()
rank = comm.Get_rank()

if (rank == root):
  a = np.arange(N, dtype='i')
else:
  a = np.zeros(N, dtype='i')

# broadcast array to all processes
comm.Bcast([a,MPI.INT],root=root)
	
# "suboptimal" partitioning
chunk = math.ceil(float(N) / numtasks);

for i in range(rank*chunk,min(N,(rank+1)*chunk)):
    result[0] += a[i] * a[i]; n+=1

print("process",rank,"chunk size:",n)

# collect results & sum them up
comm.Reduce(result,final_result,
             op=MPI.SUM,root=root)

if (rank ==  root):
  print("result=",final_result)

File: Lectures & Talks/Day 3/Examples/mpi_collective.c
#include <mpi.h>
#include <stdio.h>
#include <math.h>

#define N 100
#define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))

int main(int argc, char *argv[]) {

int numtasks, rank, n=0, i, root=0, chunk;
int a[N], b[N], result=0, final_result;

MPI_Init(&argc,&argv);
MPI_Comm_size(MPI_COMM_WORLD,&numtasks);
MPI_Comm_rank(MPI_COMM_WORLD,&rank);

if (rank == root)
  for (i=0; i<N; i++) {a[i] = i;}
// broadcast array to all processes
MPI_Bcast(a,N,MPI_INT,root,MPI_COMM_WORLD);

// "suboptimal" partitioning
chunk = ceil((float) N / numtasks);

for (i=rank*chunk;i<MIN(N,(rank+1)*chunk);i++) {
   result += a[i] * a[i]; n++;
}

printf("process %d chunk size: %d\n",rank,n);
// collect results & sum them up
MPI_Reduce(&result,&final_result,1,MPI_INT,
	     MPI_SUM,root,MPI_COMM_WORLD);

if (rank ==  root) {
   printf("result= %d\n",final_result);
}

MPI_Finalize();
}
